File: app/src/main/java/ru/nikshlykov/englishwordsapp/WordDao.java
package ru.nikshlykov.englishwordsapp;


import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface WordDao {

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long insert(Word word);

    @Update
    int update(Word word);

    /*// Будет использован при сбрасывании прогресса всей подгруппы.
    // НЕОБХОДИМО УЧЕСТЬ ТО, ЧТО ОБНОВЛЯЮТСЯ ВСЕ ПОЛЯ ИЗ ENTITY.????
    @Update
    int updateMultiple(List<Word> words);*/

    @Delete
    int delete(Word word);

    /*// Можно использовать при удалении своей группы, слова в которой связаны только с ней.
    // Это будут созданные пользователем слова.
    @Delete
    int deleteMultiple(List<Word> words);*/

    @Query("SELECT * FROM Words WHERE _id = :wordID")
    Word getWordById(long wordID);

    @Query("SELECT Words.* FROM Words, Links WHERE Words._id = Links.WordId and Links.SubgroupId = :subgroupId")
    LiveData<List<Word>> getWordsFromSubgroup(long subgroupId);

    @Query("SELECT * FROM Words ORDER BY _id LIMIT 1")
    Word getWordWithMinId();

    @Query("SELECT Words.* FROM Words, Links, Subgroups WHERE Words._id = Links.WordId AND Subgroups._id = Links.SubgroupId AND Subgroups.IsStudied = 1")
    Word[] getAllWordsFromStudiedSubgrops();

    @Query("SELECT Words.* FROM Words, Links, Subgroups WHERE Words._id = Links.WordId AND Subgroups._id = Links.SubgroupId AND Subgroups.IsStudied = 1")
    LiveData<List<Word>> getWords();
}

File: core-db/src/main/java/ru/nikshlykov/data/database/daos/WordDao.kt
package ru.nikshlykov.data.database.daos

import androidx.room.*
import kotlinx.coroutines.flow.Flow
import ru.nikshlykov.data.database.models.Word

@Dao
interface WordDao {

  @Insert(onConflict = OnConflictStrategy.IGNORE)
  suspend fun insert(word: Word): Long

  @Update
  suspend fun update(word: Word): Int

  @Query("SELECT * FROM Words WHERE _id = :wordId")
  suspend fun getWordById(wordId: Long): Word

  @Query(
    "SELECT Words.* FROM Words, Links " +
            "WHERE Words._id = Links.WordId and Links.SubgroupId = :subgroupId " +
            "ORDER BY Words.LearnProgress DESC, Words.word"
  )
  fun getWordsFromSubgroupByProgressFlow(subgroupId: Long): Flow<List<Word>>

  @Query(
    "SELECT Words.* FROM Words, Links " +
            "WHERE Words._id = Links.WordId and Links.SubgroupId = :subgroupId " +
            "ORDER BY Words.Word, Words.LearnProgress DESC"
  )
  fun getWordsFromSubgroupByAlphabetFlow(subgroupId: Long): Flow<List<Word>>

  @Query("SELECT * FROM Words ORDER BY _id LIMIT 1")
  suspend fun wordWithMinId(): Word

  @Query(
    "SELECT DISTINCT Words.* FROM Words, Links, Subgroups " +
            "WHERE Words._id = Links.WordId AND Subgroups._id = Links.SubgroupId " +
            "AND Subgroups.IsStudied = 1 ORDER BY Words.LearnProgress DESC, Words.Priority"
  )
  suspend fun getWordsFromStudiedSubgroups(): List<Word>

  @Query(
    "SELECT DISTINCT Words.* FROM Words, Links, Subgroups " +
            "WHERE Words._id = Links.WordId AND Subgroups._id = Links.SubgroupId " +
            "AND Subgroups.IsStudied = 1 AND Words.LearnProgress >= 0 ORDER BY Words.LearnProgress DESC, Words.Priority"
  )
  suspend fun getNotNewWordsFromStudiedSubgroups(): List<Word>

  @Query("UPDATE Words SET LearnProgress = -1 WHERE Words._id = :wordId")
  suspend fun resetWordProgress(wordId: Long): Int

  @Query("UPDATE Words SET LearnProgress = -1 WHERE Words._id in (SELECT w._id FROM Subgroups s INNER JOIN Links l on l.SubgroupId == s._id INNER JOIN Words w on l.WordId == w._id WHERE s._id == :subgroupId)")
  suspend fun resetWordsProgressFromSubgroup(subgroupId: Long): Int
}